hudi-common/src/main/java/org/apache/hudi/io/storage/HoodieHFileReader.java

```diff
@@ -209,12 +212,23 @@ public R next() {
 
   @Override
   public Option getRecordByKey(String key, Schema readerSchema) throws IOException {
-    HFileScanner scanner = reader.getScanner(false, true);
+    byte[] value = null;
     KeyValue kv = new KeyValue(key.getBytes(), null, null, null);
-    if (scanner.seekTo(kv) == 0) {
-      Cell c = scanner.getKeyValue();
-      byte[] keyBytes = Arrays.copyOfRange(c.getRowArray(), c.getRowOffset(), c.getRowOffset() + c.getRowLength());
-      R record = getRecordFromCell(c, getSchema(), readerSchema);
+
+    synchronized (this) {
```

Comment:
can you help me understand how getRecordByKey() will be called concurrently? in other words, can a single reader be called by multiple callers concurrently. from my understanding, a single reader won't trigger concurrent getRecordByKey() calls.

Reply:
The concurrent calls happen when TimelineService is being used. TimelineService is based on Javalin/Jetty which has a thread-per-request synchronous model. Multiple threads in parallel will call in the app's router to handle the HTTP request. 

Within the org.apache.hudi.timeline.service.FileSystemViewHandler, we handle the remote calls by code similar to -- viewManager.getFileSystemView(basePath).getLatestXXX(...)

When Metadata Table is enabled, the file system view should be HoodieMetadataFileSystemView which uses a HoodieTableMetadata (class variable). Hence, the HoodieTableMetadata.getAllFilesInPartition() will be called concurrently on multiple threads. This function internally calls the getRecordByKey().

Reply:
got it, thanks for the info.